File: emotion-ai/api/typescript/src/stream.ts
/**
 * WIA Emotion AI WebSocket Streaming
 *
 * Real-time emotion streaming client.
 */

import { EmotionEvent } from './models';

export interface StreamConfig {
    apiKey: string;
    url: string;
    modalities: string[];
    frameRate: number;
    detectActionUnits?: boolean;
    detectMicroExpressions?: boolean;
    confidenceThreshold?: number;
}

export type StreamEventType =
    | 'emotion'
    | 'aggregate'
    | 'error'
    | 'connected'
    | 'disconnected';

export interface EmotionStreamEvent {
    type: 'emotion.event';
    eventId: string;
    timestamp: string;
    sequence: number;
    latencyMs: number;
    emotions: Array<{
        category: string;
        intensity: number;
        confidence: number;
    }>;
    actionUnits?: Array<{
        au: string;
        intensity: number;
        confidence: number;
    }>;
    dimensional?: {
        valence: number;
        arousal: number;
    };
}

export interface AggregateEvent {
    type: 'emotion.aggregate';
    timestamp: string;
    windowStart: string;
    windowEnd: string;
    frameCount: number;
    aggregate: {
        dominantEmotion: string;
        emotionDistribution: Record<string, number>;
        averageValence: number;
        averageArousal: number;
        emotionalStability: number;
        engagementScore: number;
    };
}

type EventHandler<T> = (event: T) => void;

/**
 * Real-time emotion streaming client using WebSocket.
 *
 * @example
 * ```typescript
 * const stream = client.createStream({ modalities: ['facial', 'voice'] });
 *
 * stream.on('emotion', (event) => {
 *     console.log(`Emotion: ${event.emotions[0].category}`);
 * });
 *
 * stream.on('aggregate', (summary) => {
 *     console.log(`Dominant: ${summary.aggregate.dominantEmotion}`);
 * });
 *
 * await stream.start();
 *
 * // Send frames
 * stream.sendFrame(videoFrame);
 * stream.sendAudio(audioChunk);
 *
 * // Stop
 * await stream.stop();
 * ```
 */
export class EmotionStream {
    private config: StreamConfig;
    private handlers: Map<StreamEventType, EventHandler<unknown>[]> = new Map();
    private sessionId: string | null = null;
    private connected = false;
    private sequence = 0;

    constructor(config: StreamConfig) {
        this.config = {
            detectActionUnits: true,
            detectMicroExpressions: false,
            confidenceThreshold: 0.5,
            ...config
        };

        // Initialize handler arrays
        const eventTypes: StreamEventType[] = [
            'emotion', 'aggregate', 'error', 'connected', 'disconnected'
        ];
        eventTypes.forEach(type => this.handlers.set(type, []));
    }

    /**
     * Register event handler.
     */
    on(eventType: 'emotion', handler: EventHandler<EmotionStreamEvent>): this;
    on(eventType: 'aggregate', handler: EventHandler<AggregateEvent>): this;
    on(eventType: 'error', handler: EventHandler<Error>): this;
    on(eventType: 'connected' | 'disconnected', handler: EventHandler<void>): this;
    on(eventType: StreamEventType, handler: EventHandler<unknown>): this {
        const handlers = this.handlers.get(eventType);
        if (handlers) {
            handlers.push(handler);
        }
        return this;
    }

    /**
     * Start the streaming session.
     */
    async start(): Promise<void> {
        throw new Error('Install full SDK: npm install @wia/emotion-ai');
    }

    /**
     * Stop the streaming session.
     */
    async stop(requestSummary = true): Promise<Record<string, unknown> | undefined> {
        throw new Error('Install full SDK: npm install @wia/emotion-ai');
    }

    /**
     * Send a video frame for analysis.
     */
    sendFrame(frame: Buffer | Uint8Array, format: 'jpeg' | 'png' = 'jpeg'): void {
        throw new Error('Install full SDK: npm install @wia/emotion-ai');
    }

    /**
     * Send an audio chunk for analysis.
     */
    sendAudio(audio: Buffer | Uint8Array, encoding = 'LINEAR16'): void {
        throw new Error('Install full SDK: npm install @wia/emotion-ai');
    }

    /**
     * Pause the streaming session.
     */
    pause(): void {
        throw new Error('Install full SDK: npm install @wia/emotion-ai');
    }

    /**
     * Resume the streaming session.
     */
    resume(): void {
        throw new Error('Install full SDK: npm install @wia/emotion-ai');
    }

    /**
     * Check if stream is connected.
     */
    get isConnected(): boolean {
        return this.connected;
    }

    /**
     * Get current session ID.
     */
    getSessionId(): string | null {
        return this.sessionId;
    }
}

File: emotion-ai/api/python/wia_emotion_ai/stream.py
"""
WIA Emotion AI WebSocket Streaming

Real-time emotion streaming client.
"""

from typing import Optional, List, Callable, Dict, Any
from dataclasses import dataclass
import json


@dataclass
class StreamConfig:
    """Streaming session configuration."""
    modalities: List[str]
    frame_rate: int = 15
    detect_action_units: bool = True
    detect_micro_expressions: bool = False
    confidence_threshold: float = 0.5
    emit_interval_ms: int = 100


class EmotionStream:
    """
    Real-time emotion streaming client using WebSocket.

    Example:
        stream = client.create_stream(modalities=["facial", "voice"])

        @stream.on("emotion")
        def handle_emotion(event):
            print(f"Emotion: {event['emotions'][0]['category']}")

        @stream.on("aggregate")
        def handle_aggregate(summary):
            print(f"Dominant: {summary['dominant_emotion']}")

        stream.start()

        # Send frames
        stream.send_frame(video_frame)
        stream.send_audio(audio_chunk)

        # Stop
        stream.stop()
    """

    def __init__(
        self,
        api_key: str,
        base_url: str,
        modalities: List[str],
        frame_rate: int = 15
    ):
        """
        Initialize streaming client.

        Args:
            api_key: WIA API key
            base_url: WebSocket URL
            modalities: List of modalities to stream
            frame_rate: Target frame rate
        """
        self.api_key = api_key
        self.base_url = base_url
        self.config = StreamConfig(
            modalities=modalities,
            frame_rate=frame_rate
        )
        self._handlers: Dict[str, List[Callable]] = {
            "emotion": [],
            "aggregate": [],
            "error": [],
            "connected": [],
            "disconnected": []
        }
        self._session_id: Optional[str] = None
        self._connected = False
        self._sequence = 0

    def on(self, event_type: str) -> Callable:
        """
        Decorator to register event handler.

        Args:
            event_type: Event type (emotion, aggregate, error, etc.)

        Example:
            @stream.on("emotion")
            def handle_emotion(event):
                print(event)
        """
        def decorator(func: Callable) -> Callable:
            if event_type not in self._handlers:
                self._handlers[event_type] = []
            self._handlers[event_type].append(func)
            return func
        return decorator

    def start(self) -> None:
        """
        Start the streaming session.

        Connects to WebSocket and sends session.start message.
        """
        # Implementation would use websockets library
        raise NotImplementedError("Install full SDK: pip install wia-emotion-ai")

    def stop(self, request_summary: bool = True) -> Optional[Dict[str, Any]]:
        """
        Stop the streaming session.

        Args:
            request_summary: Request session summary

        Returns:
            Session summary if requested
        """
        raise NotImplementedError("Install full SDK: pip install wia-emotion-ai")

    def send_frame(self, frame: bytes, format: str = "jpeg") -> None:
        """
        Send a video frame for analysis.

        Args:
            frame: Video frame data (JPEG/PNG bytes)
            format: Frame format (jpeg, png)
        """
        raise NotImplementedError("Install full SDK: pip install wia-emotion-ai")

    def send_audio(self, audio: bytes, encoding: str = "LINEAR16") -> None:
        """
        Send an audio chunk for analysis.

        Args:
            audio: Audio data
            encoding: Audio encoding (LINEAR16, FLAC, etc.)
        """
        raise NotImplementedError("Install full SDK: pip install wia-emotion-ai")

    def pause(self) -> None:
        """Pause the streaming session."""
        raise NotImplementedError("Install full SDK: pip install wia-emotion-ai")

    def resume(self) -> None:
        """Resume the streaming session."""
        raise NotImplementedError("Install full SDK: pip install wia-emotion-ai")

    @property
    def is_connected(self) -> bool:
        """Check if stream is connected."""
        return self._connected

    @property
    def session_id(self) -> Optional[str]:
        """Get current session ID."""
        return self._session_id


def create_session_start_message(config: StreamConfig) -> Dict[str, Any]:
    """Create session.start message."""
    return {
        "type": "session.start",
        "config": {
            "modalities": config.modalities,
            "facial_config": {
                "frame_rate": config.frame_rate,
                "detect_action_units": config.detect_action_units,
                "detect_micro_expressions": config.detect_micro_expressions
            },
            "output_config": {
                "confidence_threshold": config.confidence_threshold,
                "emit_interval_ms": config.emit_interval_ms
            }
        }
    }
